File: 844.backspace-string-compare.scala
/*
 * @lc app=leetcode id=844 lang=scala
 *
 * [844] Backspace String Compare
 */

// @lc code=start
object Solution {
    def compact(s: String): String =
        s.foldLeft(List.empty[Char]) {
            case (acc, '#') => acc.drop(1)
            case (acc, elem) => elem :: acc
        }.mkString

    def backspaceCompare(s: String, t: String): Boolean = {
        compact(s) == compact(t)
    }
}
// @lc code=end

File: 844.backspace-string-compare.py
#
# @lc app=leetcode id=844 lang=python3
#
# [844] Backspace String Compare
#

# @lc code=start
class Solution:
    def backspaceCompare(self, s: str, t: str) -> bool:
        i = 0
        xs = list(s)
        while i < len(xs):
            if xs[i] == '#':
                xs.pop(i)
                if i > 0:
                    xs.pop(i - 1)
                    i -= 1
            else:
                i += 1

        i = 0
        xt = list(t)
        while i < len(xt):
            if xt[i]  == '#':
                xt.pop(i)
                if i > 0:
                    xt.pop(i - 1)
                    i -= 1
            else:
                i += 1

        return xs == xt
# @lc code=end
